shared/charts/gpii-flushtokens/templates/cronjob.yaml

```diff
@@ -0,0 +1,58 @@
+apiVersion: batch/v1beta1
+kind: CronJob
+metadata:
+  namespace: {{ .Release.Namespace | quote }}
+  name: {{ template "flushtokens.name" . }}
+spec:
+  schedule: "*/15 * * * *"
+  concurrencyPolicy: Allow
+  startingDeadlineSeconds: 200
+  jobTemplate:
+    spec:
+      template:
+        metadata:
+          labels:
+            app: {{ template "flushtokens.name" . }}
+        spec:
+          shareProcessNamespace: true
+          containers:
+          - name: gpii-flushtokens
+            image: "{{ .Values.image.repository }}@{{ .Values.image.checksum }}"
+            command: [ "/bin/sh", "-c", "-x", "/app/scripts/dockerDeleteExpiredAccessTokens.sh" ]
+            env:
+            - name: GPII_COUCHDB_URL
+              value: '{{ .Values.couchdbUrl }}'
+            # Use '--deleteAll' to delete all of the access tokens
+            - name: DELETE_ALL
+              value: 'false'
+          - name: istio-proxy-manager
+            image: gcr.io/gke-release/istio/proxyv2:1.0.6-gke.3@sha256:1f6dac3dbf69c0b0d519398003b6572e0218f695f1be295a2ffff95df2395ed0
+            securityContext:
+              allowPrivilegeEscalation: false
+              runAsUser: 1337
+            command: [ "/bin/sh", "-c" ]
+            args:
+              - for i in `seq 1 30`;
```

Comment:
I wanted to avoid a potential infinite loop.  If `dockerDeleteExpiredAccessTokens.sh` is never detected, then the while loop will never exit.  After experimenting with this code and different sleep time-outs, the script is either detected immediately or on the second try, so 30 is probably overkill.

That said, is it safe to use a `while` loop?  You know more about clusters and how they start up better than I, and if you think a `while` loop is good enough, then I can change it.

I will change for now just to check that it does work.

Reply:
I think we would benefit from the simpler logic here.  I guess there's some chance we would miss the node -> ie. it would be so fast that it would start & finish in between those 2 seconds.., or the script would fail before it ever starts the node.

And I have played a bit with the available options in CronJob spec, and I would propose to solve this issue using the `concurrencyPolicy: Replace` (sorry I know before I suggested to change it to `Forbid` :))

With the following spec I was able to get behaviour that I think is desired:
```
spec:
  schedule: "*/1 * * * *"
  successfulJobsHistoryLimit: 5
  failedJobsHistoryLimit: 1
  concurrencyPolicy: Replace
...
```
- if the pod runs and ends with success, incl. killing the istio-proxy - `successfulJobsHistoryLimit` (in the example 5) number of completely stopped pods is kept around
- if the pod fails, no matter whether the istio-proxy kill code succeeds or not, it will be restarted until it's time for next run (1m in the example) and then it's replaced with the new pod
- if the pod runs and ends with success, but without killing the istio-proxy, it's again going to be replaced with the new pod

So this way there's never more than one pod running. The only downside I can see is if the code would actually take more time than the cron interval, then we would kill it in the middle - but I don't think it's an issue... or?

Reply:
FYI, here is the log of one of the runs today:
```
2019-07-05 12:30:11.857 EDT dockerDeleteExpiredAccessTokens.sh started.
2019-07-05 12:30:11.858 EDT Waiting for dockerDeleteExpiredAccessTokens.sh to finish...
2019-07-05 12:30:13.861 EDT Waiting for dockerDeleteExpiredAccessTokens.sh to finish...
2019-07-05 12:30:15.863 EDT Waiting for dockerDeleteExpiredAccessTokens.sh to finish...
2019-07-05 12:30:17.865 EDT dockerDeleteExpiredAccessTokens.sh finished.
2019-07-05 12:30:17.865 EDT Sending TERM to pilot-agent
```
The deletion script took about 4-6 sec start-to-finish.  Usually it's 2 seconds; typically there are only two `Waiting for dockerDeleteExpiredAccessTokens.sh to finish...` messages.